### ISSUE TITLE
Incorrect field values on client side after updateOne() with aggregation pipeline

### ISSUE THREAD

**Issue by Torgen** [meteor/meteor#11914]
Issue appears in 2.6
Issue did not appear in 2.5.6
Ubuntu 20.04 under WSL2 on Windows 10

Using the updateOne() method on rawCollection, passing an aggregation pipeline, I reorder the elements of an array field.
Expected behavior: Minimongo on the client initially contains the original order, and after sync it contains the modified order.
Actual behavior: Minimongo on the client contains an array that is neither the original nor final order; it has a duplicate element and is missing another element.

Repro: I saw this in my app during an upgrade. A version with the upgrade and some extra debugging logging is at https://github.com/Torgen/codex-blackboard/commit/1206e6ed44078e7b3dc4857673a2b1be694f379c; you can see the output in the github actions coverage.full.lcov run. The first test case ("drag-and-drop" > "allows drag and drop within a meta") shows the issue. For that run, the log messages for both the sim and real show that after the mutation, the puzzles field contains
```
["9yPbdgzeNXEBKMyoL","QnpLAx4aLJtF2w2qu","JTtwWmMi6gTrqrKtj","D79wFGQJ8fko3YzfQ","3HibG7MqfvLCyppw5","nwmoR5N3mYWA3sXtC","5vxGCadHmFKsqbNk6","GMqSk6h7qu6TQF6id"]
```
but at some point on the client, the puzzles field contains 
```
["9yPbdgzeNXEBKMyoL","QnpLAx4aLJtF2w2qu","JTtwWmMi6gTrqrKtj","D79wFGQJ8fko3YzfQ","3HibG7MqfvLCyppw5","nwmoR5N3mYWA3sXtC","5vxGCadHmFKsqbNk6","5vxGCadHmFKsqbNk6"]
```
(i.e. the last two elements are both "5vxGCadHmFKsqbNk6" and "GMqSk6h7qu6TQF6id" is missing). If you run the app tests yourself the values will differ because of the random ID generation, but I see the problem consistently.
Since I had only recently made the change to use the aggregation pipeline to make the change atomically, I was able to revert it and go back to the double checked logic that does more database operations, which is https://github.com/Torgen/codex-blackboard/commit/bef7eec80721a7cd6e9f207eafd35519feef5c44. The tests pass at this commit, which shows that the difference is using the aggregation pipeline in updateOne().

**Comment by renanccastro** [comment]
Hi @Torgen thanks for reporting
Would it be possible for you to wrap up a smaller reproduction repo?
Something simple, like a new meteor project without coffeescript with just this case.

It would really make it easier for us to repro and fix.

**Comment by tjramage** [comment]
Hi Meteor team! Just to flag, we are also seeing very unusual data come back via DDP "changed" messages in v2.6. This data does not reflect what is in the database and it doesn't even resemble the correct shape, which is highly unusual.

See screenshot below of an object that has dynamic keys, and each key should have it's own array of strings. However, we see two keys have suddenly changed into an object with `{ a: true, u1: METEOR_DOC_ID }`:

<img width="1940" alt="Screenshot 2022-02-15 at 14 13 52" src="https://user-images.githubusercontent.com/5517057/154079265-cd89b9f5-8373-419b-81cf-c4935f688252.png">

Upon refresh, the data is correct and resembles the same as the database (an array of strings)...

**Comment by renanccastro** [comment]
@tjramage do you have any hints on what type of updates are spawning those wrong changed messages?

Mongo 5 updated the oplog implementation format, so we are reverse-engineering the entries. Some are probably not following the pattern, or are edge cases like the aggregation pipeline of OP or even your case.

**Comment by tjramage** [comment]
@renanccastro sure thing – this is the exact query. The odd thing is that it doesn't fail all the time. We can get Minimongo to return this unusual data when running 100's of calls to the method on the server-side.

Query:

```
MyCollection.update(documentId, {
	$push: {
		[`my.nested.path.${keyId}`]: {
			$each: [referenceId],
			$position: index
		}
	}
})
```

The above query is essentially inserting the `referenceId` string at the specified `index` into the array.

**Comment by tjramage** [comment]
Awesome, thanks so much @renanccastro! Please do let me know if there's any other information I can provide 👍

**Comment by renanccastro** [comment]
Hi @tjramage 
I couldn't repro in a simple project.
Inside this document:

```js
       collection.insert({
            "test":{
                "testNested": [
                    '1',
                    '2',
                    '3'
                ]
            }
        })
```
I'm updating it with:

```js
        collection.update({}, {$push: {[`test.testNested`]: {$each: ["5"], $position: 2}}})
```

And I don't get the behavior you are seeing.

**Comment by tjramage** [comment]
@renanccastro – indeed, it's an odd bug. The issue happens when we run an import from an Excel file upload, in which the Meteor method can be called 1000's of times. Also, are you analysing every single DDP changed message? As the data corrects itself in subsequent messages...

**Comment by tjramage** [comment]
Just to doubel check, @renanccastro, do you have any insight into the object structure of `{ a: true, u1: <VALUE HERE> }`? Does this actually look like a possible Mongo oplog sync issue?

**Comment by Torgen** [comment]
Made a minimal app at https://github.com/Torgen/meteor-11914-repro that has the same problem. Use `meteor npm test` to run the app tests. In the `meteor-2.5.6` branch, the list has the expected contents. In `meteor-2.6` the element of the list that I move gets duplicated. I still have the logging in the sim and server versions so you can see that what's in the database is correct, but what gets delivered over the subscription is wrong.

**Comment by renanccastro** [comment]
thanks very much @Torgen, I'm analyzing this right now.

@tjramage I've tested doing a update per second during some minutes, all changed messages seemed fine.
That structure does seem like an issue with oplog conversion, as this kind of object is the "array" inserter of oplog v2, but it's not well structured. Are there perhaps any other update in the same method? that could be a clue.

**Comment by renanccastro** [comment]
@Torgen https://github.com/Torgen/meteor-11914-repro is empty. Is there perhaps any unpushed changes in your git?

**Comment by renanccastro** [comment]
Now I see the branches :man_facepalming: sorry.

**Comment by renanccastro** [comment]
@Torgen this PR fixes your issue: https://github.com/meteor/meteor/pull/11929
Can you confirm that it works?

@tjramage it might fix yours too. Can you please give it a shot?

**Comment by Torgen** [comment]
Are there instructions somewhere for running my app with meteor tool from a git clone?

**Comment by tjramage** [comment]
> Are there instructions somewhere for running my app with meteor tool from a git clone?

@Torgen – you can pull down the repo from @renanccastro's PR branch and then copy the `/packages/mongo` folder into your own `/packages` folder in the root of your Meteor project.

**Comment by renanccastro** [comment]
@tjramage do you have an entry I can check in full? that would be the offending "oplogEntry" that causes this in the logs.

**Comment by renanccastro** [comment]
@Torgen another way is to clone Meteor in a folder, checking out the branch [fix/oplog-aggregation](https://github.com/meteor/meteor/tree/fix/oplog-aggregation), and run your app using the binary "meteor" in the folder.

you should use:

/my/path/to/repo/meteor run test

**Comment by tjramage** [comment]
> @tjramage do you have an entry I can check in full? that would be the offending "oplogEntry" that causes this in the logs.

@renanccastro sure thing:

```
{
  oplogEntry: {
    slayout: {
      sjourneyStepIds: {
        sj4aqp3tiK6xCPCYu8: {
          a: true,
          u2: 'zTkxivNrKuBi2iJ2m'
        }
      }
    }
  }
}
```

Notice that the sField for `sj4aqp3tiK6xCPCYu8` ends with `u8` – this is triggering a false positive...

**Comment by tjramage** [comment]
@renanccastro – I have a pretty simple reproduction of this issue now. I assume it's not needed with the above information, but just shout if you would find that helpful and I'll share a repo 👍

**Comment by renanccastro** [comment]
hi, @tjramage I've pushed a fix to your issue in that same branch.

Can you please let me know if it works? thanks.

**Comment by Torgen** [comment]
Tried it, but got
```
Error: Cannot find module '@wry/context'
```
Tried running `npm install` in the meteor repo root directory but it didn't help.

**Comment by tjramage** [comment]
@renanccastro – perfect! All working nicely now 👏

**Comment by renanccastro** [comment]
I've just released mongo@1.14.5 as an independent release.
You can do a meteor update mongo to get the latest.

Thanks for your help.

**Comment by tjramage** [comment]
Happy to help – appreciate the fast turnaround also! 👍
